### PR TITLE
mgr/orchestrator: validate lists in spec jsons

### PR DESCRIPTION
**Pull request by Daniel-Pivonka**
validate that expected lists in specs are actually lists in spec json
allow a single string to be given by converting it into a list containing that string
return error for all other unexpected types

Fixes: https://tracker.ceph.com/issues/50102
Signed-off-by: Daniel Pivonka <dpivonka@redhat.com>

### CODE REVIEW

File: src/python-common/ceph/deployment/service_spec.py [hunk 1]
```diff
@@ -556,6 +558,17 @@ def from_json(cls: Type[ServiceSpecT], json_spec: Dict) -> ServiceSpecT:
 
         return _cls._from_json_impl(c)  # type: ignore
 
+    @staticmethod
+    def validate_json(json_spec: dict) -> dict:
+        if 'networks' in json_spec:
+            if type(json_spec['networks']) is not list:
```
Comment (tchaikov) [inline src/python-common/ceph/deployment/service_spec.py:564]: why not use `isinstance()`?

File: src/python-common/ceph/deployment/hostspec.py [hunk 2]
```diff
@@ -1,9 +1,22 @@
+import errno
 try:
     from typing import Optional, List, Any
 except ImportError:
     pass  # just for type checking
 
 
+class HostSpecValidationError(Exception):
```
Comment (sebastian-philipp) [inline src/python-common/ceph/deployment/hostspec.py:8]:
```suggestion
class SpecValidationError(Exception):
```
and just use it everywhere? I doubt anyone needs to differentiate them.

**Review by jmolmo — APPROVED** [review]

File: src/python-common/ceph/deployment/service_spec.py [hunk 3]
```diff
@@ -556,6 +545,17 @@ def from_json(cls: Type[ServiceSpecT], json_spec: Dict) -> ServiceSpecT:
 
         return _cls._from_json_impl(c)  # type: ignore
 
+    @staticmethod
+    def validate_json(json_spec: dict) -> dict:
```
Comment (tchaikov) [inline src/python-common/ceph/deployment/service_spec.py:549]: this function not only validate `json_spec`, it also normalizes it. probably could rename it to `normalize_json()`?

File: src/python-common/ceph/deployment/service_spec.py [hunk 4]
```diff
@@ -556,6 +545,17 @@ def from_json(cls: Type[ServiceSpecT], json_spec: Dict) -> ServiceSpecT:
...
+        if 'networks' in json_spec:
+            if not isinstance(json_spec['networks'], list):
+                tmp = json_spec['networks']
```
Comment (tchaikov) [inline src/python-common/ceph/deployment/service_spec.py:552]:
```suggestion
        networks = json_spec.get('networks')
        if networks is None:
            return json_spec
        if isinstance(networks, list):
            return json_spec
        if not isinstance(networks, str):
            raise SpecValidationError(f'Networks ({network}) must be a string or list of strings')
        json_spec['networks'] = [networks]
        return json_spec
```

**Review by tchaikov — APPROVED** [review]

**Comment by Daniel-Pivonka** [comment]
jenkins test docs

**Comment by tchaikov** [comment]
- https://pulpito.ceph.com/kchai-2021-04-15_08:31:03-rados-wip-kefu-testing-2021-04-15-1359-distro-basic-smithi/
- https://pulpito.ceph.com/kchai-2021-04-16_02:02:40-rados-wip-kefu-testing-2021-04-15-1359-distro-basic-gibba/

the valgrind issue is tracked by https://tracker.ceph.com/issues/50299

**Pull request merged**